File: PDFReaderPlugin/Helpers/PasteboardHelper.m
//
//  PasteboardHelper.m
//  excerptor
//
//  Created by Chen Guo on 17/05/2015.
//  Copyright (c) 2015 guoc. All rights reserved.
//

@import AppKit;

#import "PasteboardHelper.h"
#import "SelectionLocation.h"

@implementation PasteboardHelper

static NSString *const InputPasteboardName = @"name.guoc.excerptor.ExcerptorToPDFReader";
static NSString *const OutputPasteboardName = @"name.guoc.excerptor.PDFReaderToExcerptor";
static NSString *const PasteboardType = @"org.nspasteboard.TransientType";

+ (Location *)readPasteboard {
    NSPasteboard *pasteboard = [NSPasteboard pasteboardWithName:InputPasteboardName];
    NSData *data = [pasteboard dataForType:PasteboardType];
    if (!data) {
        return nil;
    }
    NSObject *unarchivedObject = [NSKeyedUnarchiver unarchiveObjectWithData:data];
    if (![unarchivedObject isKindOfClass:[Location class]]) {
        return nil;
    }
    return (Location *)unarchivedObject;
}

+ (bool)writePasteboardWithSelectionLocation: (SelectionLocation *)selectionLocation {
    NSData *data = [NSKeyedArchiver archivedDataWithRootObject:selectionLocation];
    NSPasteboard *pasteboard = [NSPasteboard pasteboardWithName:OutputPasteboardName];
    [pasteboard clearContents];
    return [pasteboard setData:data forType:PasteboardType];
}

+ (bool)writePasteboardWithErrorMessage: (NSString *)errorMessage {
    if ([[PasteboardHelper readPasteboard] isKindOfClass:[Location class]]) {
        return NO;
    }
    NSData *data = [NSKeyedArchiver archivedDataWithRootObject: [@"Excerptor PDFReaderPlugin: " stringByAppendingString: errorMessage]];
    NSPasteboard *pasteboard = [NSPasteboard pasteboardWithName:OutputPasteboardName];
    [pasteboard clearContents];
    return [pasteboard setData:data forType:PasteboardType];
}

+ (void)clearInputPasteboard {
    NSPasteboard *pasteboard = [NSPasteboard pasteboardWithName:InputPasteboardName];
    [pasteboard clearContents];
}

@end

File: Excerptor/Helpers/PasteboardHelper.swift
//
//  PasteboardHelper.swift
//  excerptor
//
//  Created by Chen Guo on 17/05/2015.
//  Copyright (c) 2015 guoc. All rights reserved.
//

import Cocoa

class PasteboardHelper {

    struct Constants {
        static let OutputPasteboardName = "name.guoc.excerptor.ExcerptorToPDFReader"
        static let InputPasteboardName = "name.guoc.excerptor.PDFReaderToExcerptor"
        static let PasteboardType = "org.nspasteboard.TransientType"
        static let ExcerptorToPDFReaderHeader = "ExcerptorToPDFReader:"
        static let PDFReaderToExcerptorHeader = "PDFReaderToExcerptor:"
    }

    class func writeGeneralPasteboardWith(selectionLink: SelectionLink?) -> Bool {
        if let selectionLink = selectionLink {
            let pasteboard = NSPasteboard.generalPasteboard()
            pasteboard.clearContents()
            let attributedString = NSAttributedString(string: selectionLink.selectionTextArrayInOneLine, attributes: [NSLinkAttributeName: selectionLink.string])
            return pasteboard.writeObjects([attributedString])
                && pasteboard.setString(selectionLink.string, forType:NSPasteboardTypeString)
        }
        return false
    }

    class func readExcerptorPasteboard() -> Location! {
        let pasteboard = NSPasteboard(name: Constants.InputPasteboardName)
        guard let data = pasteboard.dataForType(Constants.PasteboardType) else {
            exitWithError("Could not get selection/annotation information")
        }
        guard let unarchivedObject = NSKeyedUnarchiver.unarchiveObjectWithData(data) else {
            exitWithError("Could not unarchive object with data " + data.description)
        }
        guard let location = unarchivedObject as? Location else {
            if let errorMessage = unarchivedObject as? NSString {
                exitWithError("Error message in pasteboard: " + String(errorMessage))
            } else {
                exitWithError("Could not convert unarchived object to location " + unarchivedObject.description)
            }
        }
        return location
    }

    class func writeExcerptorPasteboardWithLocation(location: Location) -> Bool {
        let pasteboard = NSPasteboard(name: Constants.OutputPasteboardName)
        pasteboard.clearContents()
        let data = NSKeyedArchiver.archivedDataWithRootObject(location)
        return pasteboard.setData(data, forType:Constants.PasteboardType)
    }


}
